### PR TITLE
LEO-126: Allow users to provide a script that runs in their jupyter docker

### PR DESCRIPTION
**Pull request by ansingh7115**
Add a user provided script to the cluster request

Companion UI PR: https://github.com/broadinstitute/firecloud-ui/pull/1156

Both branches are built on fiab-ansingh-light-hare 35.225.55.252

Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get two thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; you can delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala [hunk 1]
```diff
@@ -40,8 +40,8 @@ case class ClusterAlreadyExistsException(googleProject: GoogleProject, clusterNa
 case class InitializationFileException(googleProject: GoogleProject, clusterName: ClusterName, errorMessage: String)
   extends LeoException(s"Unable to process initialization files for ${googleProject.value}/${clusterName.string}. Returned message: $errorMessage", StatusCodes.Conflict)
 
-case class JupyterExtensionException(gcsUri: GcsPath)
-  extends LeoException(s"Jupyter extension URI is invalid or unparseable: ${gcsUri.toUri}", StatusCodes.BadRequest)
+case class BucketObjectException(gcsUri: GcsPath)
+  extends LeoException(s"The provided URI is invalid or unparseable: ${gcsUri.toUri}", StatusCodes.BadRequest)
```
Comment (ansingh7115) [inline src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: I made the error message more generic - is that alright?

File: src/main/resources/jupyter/init-actions.sh [hunk 2]
```diff
@@ -109,6 +110,17 @@ if [[ "${ROLE}" == 'Master' ]]; then
...
     fi
 
+
+    # If a Jupyter user script was specified, copy it into the jupyter docker container.
+    if [ ! -z ${JUPYTER_USER_SCRIPT_URI} ] ; then
+      gsutil cp ${JUPYTER_USER_SCRIPT_URI} /etc
+      JUPYTER_USER_SCRIPT_ARCHIVE=`basename ${JUPYTER_USER_SCRIPT_URI}`
+      docker cp /etc/${JUPYTER_USER_SCRIPT_ARCHIVE} ${JUPYTER_SERVER_NAME}:${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT_ARCHIVE}
+      docker exec -u root -d ${JUPYTER_SERVER_NAME} chmod +x ${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT_ARCHIVE}
+      docker exec -u root -d ${JUPYTER_SERVER_NAME} ${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT_ARCHIVE}
```
Comment (helgridly) [inline src/main/resources/jupyter/init-actions.sh]: These are going to run after the Docker entry point, so pyspark will already be running, and potentially installed things won't be picked up (they definitely won't if they're Python packages).

I think the approach you need to take is to edit the dockerfile so that instead of calling pyspark, it calls a bash script that first runs the user script and then launches pyspark. You will likely have to mess around with when the dockerfile changes user to make sure it runs as root but the Jupyter user doesn't.

File: src/main/resources/swagger/api-docs.yaml [hunk 3]
```diff
@@ -501,6 +501,11 @@ definitions:
           The archive must be in tar.gz format, must not include a parent directory, and
           must have an entry point named 'main'. For more information on notebook extensions,
           see http://jupyter-notebook.readthedocs.io/en/latest/extending/frontend_extensions.html.
+      jupyterUserScriptUri:
+        type: string
+        description: |
+          Optional bucket URI to an archive containing a bash script the user wishes to run inside their jupyter docker.
+          The archive must be in .sh format and must not include a parent directory.
```
Comment (helgridly) [inline src/main/resources/swagger/api-docs.yaml]: Is it an archive or a bash script? (It should just be a batch script; no need to zip it up.)

File: src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala (hunk 1 above)
Comment (helgridly) [reply src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala]: Yeah but you might want to say GCS URI just to be clear.

File: src/main/resources/jupyter/init-actions.sh (hunk 2 above)
Comment (ansingh7115) [reply src/main/resources/jupyter/init-actions.sh]: I tested this with the following script:

`#!/usr/bin/env bash

pip install arrow`

And from my notebook I could import the library and use it - so I was able to install a python package.

But - I'm fine with doing it through the dockerfile!

**Comment by ansingh7115** [comment]
@helgridly @rtitle @vkumra-broad ready for review

File: src/main/resources/jupyter/init-actions.sh [hunk 4]
```diff
@@ -124,6 +126,17 @@ if [[ "${ROLE}" == 'Master' ]]; then
       docker exec -d ${JUPYTER_SERVER_NAME} mkdir -p ${JUPYTER_USER_HOME}/.jupyter/custom
       docker cp /etc/${JUPYTER_GOOGLE_SIGN_IN_JS} ${JUPYTER_SERVER_NAME}:${JUPYTER_USER_HOME}/.jupyter/custom/
     fi
+
+    # If a Jupyter user script was specified, copy it into the jupyter docker container.
+    if [ ! -z ${JUPYTER_USER_SCRIPT_URI} ] ; then
+      gsutil cp ${JUPYTER_USER_SCRIPT_URI} /etc
+      JUPYTER_USER_SCRIPT_ARCHIVE=`basename ${JUPYTER_USER_SCRIPT_URI}`
```
Comment (rtitle) [inline src/main/resources/jupyter/init-actions.sh]: It's not an archive, right? Maybe just name this `JUPYTER_USER_SCRIPT`

File: src/main/resources/swagger/api-docs.yaml [hunk 5]
```diff
@@ -501,6 +501,11 @@ definitions:
           The archive must be in tar.gz format, must not include a parent directory, and
           must have an entry point named 'main'. For more information on notebook extensions,
           see http://jupyter-notebook.readthedocs.io/en/latest/extending/frontend_extensions.html.
+      jupyterUserScriptUri:
+        type: string
+        description: |
+          Optional bucket URI to an archive containing a bash script the user wishes to run inside their jupyter docker.
```
Comment (rtitle) [inline src/main/resources/swagger/api-docs.yaml]: It's not an archive, can just say "bucket URI to a bash script"

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala [hunk 6]
```diff
@@ -391,7 +391,7 @@ class LeonardoServiceSpec extends TestKit(ActorSystem("leonardotest")) with Flat
     // create the cluster
     val response = leo.createCluster(defaultUserInfo, googleProject, clusterName, testClusterRequest.copy(jupyterExtensionUri = Some(jupyterExtensionUri))).failed.futureValue
```
Comment (rtitle) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala]: There's a spec in here "template a script using config values" which verifies the templating logic -- maybe add a check to make sure `jupyterUserScriptUri` gets templated.

**Comment by rtitle** [comment]
Another thought - we probably need to grant read access to this bucket (and the Jupyter extension bucket for that matter) to the pet service account or user proxy group.

@vkumra-broad faced a similar issue when creating the staging bucket in https://github.com/DataBiosphere/leonardo/pull/169. We actually added new methods to `ServiceAccountProvider` to specify the users/groups which need access to the staging bucket (in our FC implementation, we return the user proxy group retrieved from Sam). Maybe we should reuse those methods for granting access to the `userScriptURI` and `extensionURI` buckets as well.

However since that work is still in review I'm inclined to _not_ do any ACL stuff in this PR, and instead do it in a follow-on PR. Me and Vikram also need to sort out moving to `StorageDAO`. 

tldr - let's keep this PR as is, and plan to do the ACL stuff in a follow-on ticket.

**Comment by ansingh7115** [comment]
Adding automated tests for user-provided scripts will create a dependency on work @jmthibault79 is doing with creating buckets and uploading to them. Are we okay with waiting for that stuff to go in first?

**Comment by rtitle** [comment]
OK with waiting for Joel's work re: buckets.

You could always manually make a script+bucket in broad-dsde-qa for now and reference that in a test, with a TODO to auto-generate the bucket.

**Comment by jmthibault79** [comment]
Bucket PR is https://github.com/DataBiosphere/leonardo/pull/176

**Comment by vkumra-broad** [comment]
We use a public bucket for the tests, you should be able to use that for the automation test instead of creating a new bucket. 
https://github.com/DataBiosphere/leonardo/blob/861dcad9ccc2a4521df19694900e9b04cf36d563/automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/LeonardoTestUtils.scala#L29

**Comment by ansingh7115** [comment]
jenkins retest

**Comment by ansingh7115** [comment]
jenkins retest

**Comment by ansingh7115** [comment]
jenkins retest

**Comment by ansingh7115** [comment]
jenkins retest

**Comment by ansingh7115** [comment]
jenkins retest

**Comment by ansingh7115** [comment]
Leo automated tests are passing but ui tests seem broken for reasons unrelated to the UI PR D:

**Comment by ansingh7115** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala [hunk 7]
```diff
@@ -147,5 +145,35 @@ class NotebookInteractionSpec extends FreeSpec with LeonardoTestUtils with Befor
...
+        val userScriptUri = s"gs://${bucketName.value}/${userScriptObjectName.value}"
+
+        withNewBucketObject(bucketName, userScriptObjectName, userScriptString, "text/plain") { objectName =>
+          val clusterName = ClusterName("user-script-cluster" + makeRandomId())
+
+          withNewCluster(billingProject, clusterName, ClusterRequest(Map(), None, Option(userScriptUri))) { cluster =>
+            withNewNotebook(cluster) { notebookPage =>
+              notebookPage.executeCell("""print 'Hello Notebook!'""") shouldBe Some("Hello Notebook!")
+              notebookPage.executeCell("""import arrow""")
+              notebookPage.executeCell("""arrow.get(727070400)""") shouldBe Some("<Arrow [1993-01-15T04:00:00+00:00]>")
```
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/NotebookInteractionSpec.scala]: Awesome test!

File: src/main/resources/swagger/api-docs.yaml [hunk 8]
```diff
@@ -501,6 +501,11 @@ definitions:
           The archive must be in tar.gz format, must not include a parent directory, and
           must have an entry point named 'main'. For more information on notebook extensions,
           see http://jupyter-notebook.readthedocs.io/en/latest/extending/frontend_extensions.html.
+      jupyterUserScriptUri:
+        type: string
+        description: |
+          Optional bucket URI to a bash script the user wishes to run inside their jupyter docker.
+          The script must be in .sh format and must not include a parent directory.
```
Comment (rtitle) [inline src/main/resources/swagger/api-docs.yaml]: We can probably just delete this second sentence? I'm not sure it needs to be in .sh format: I could name it `my_awesome_script` and we would `chmod +x` it and try to execute it. And "must not include a parent directory" doesn't make sense for scripts. :)

File: src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala [hunk 9]
```diff
@@ -1,67 +1,47 @@
...
+import org.broadinstitute.dsde.workbench.model.google.GoogleProject
 import org.scalatest._
 import org.scalatest.concurrent.ScalaFutures
 import spray.json._
 import org.mockito.Mockito._
 import org.mockito.ArgumentMatchers.{any, eq => mockitoEq}
 import scala.concurrent.{ExecutionContext, Future}
 
-class LeonardoServiceSpec extends TestKit(ActorSystem("leonardotest")) with FlatSpecLike with Matchers with BeforeAndAfter with BeforeAndAfterAll with TestComponent with ScalaFutures with OptionValues with VCMockitoMatchers {
-  private val configFactory = ConfigFactory.load()
```
Comment (rtitle) [inline src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoServiceSpec.scala]: Oh thanks for making `LeonardoServiceSpec` use `CommonTestData`! This has been bothering me for a while.

File: src/main/resources/jupyter/init-actions.sh [hunk 10]
```diff
@@ -124,6 +126,17 @@ if [[ "${ROLE}" == 'Master' ]]; then
...
       docker cp /etc/${JUPYTER_GOOGLE_SIGN_IN_JS} ${JUPYTER_SERVER_NAME}:${JUPYTER_USER_HOME}/.jupyter/custom/
     fi
+
+    # If a Jupyter user script was specified, copy it into the jupyter docker container.
+    if [ ! -z ${JUPYTER_USER_SCRIPT_URI} ] ; then
+      gsutil cp ${JUPYTER_USER_SCRIPT_URI} /etc
+      JUPYTER_USER_SCRIPT=`basename ${JUPYTER_USER_SCRIPT_URI}`
+      docker cp /etc/${JUPYTER_USER_SCRIPT} ${JUPYTER_SERVER_NAME}:${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT}
+      docker exec -u root -d ${JUPYTER_SERVER_NAME} chmod +x ${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT}
+      docker exec -u root -d ${JUPYTER_SERVER_NAME} ${JUPYTER_HOME}/${JUPYTER_USER_SCRIPT}
```
Comment (helgridly) [inline src/main/resources/jupyter/init-actions.sh]: nice! i worried it'd be hard to get the script running as root, because we'd already have defined `USER` in the Dockerfile, but this is a nice sidestep
